crates/subspace-networking/src/protocols/request_response/request_response_factory.rs

```diff
@@ -385,13 +386,27 @@ impl RequestResponseFactoryBehaviour {
         addresses: Vec<Multiaddr>,
     ) {
         if let Some((protocol, _)) = self.protocols.get_mut(protocol_name) {
-            if protocol.is_connected(target) || connect.should_connect() {
-                let request_id = protocol.send_request(target, request, addresses);
+            if protocol.is_connected(target) {
+                let request_id = protocol.send_request(target, request);
+                let prev_req_id = self.pending_requests.insert(
+                    (protocol_name.to_string().into(), request_id).into(),
+                    (Instant::now(), pending_response),
+                );
+                debug_assert!(
+                    prev_req_id.is_none(),
+                    "Expect request id to be unique for existing peer."
+                );
+            } else if connect.should_connect() {
+                let opts = DialOpts::peer_id(*target).addresses(addresses).build();
+                let request_id = protocol.send_request(opts, request);
                 let prev_req_id = self.pending_requests.insert(
                     (protocol_name.to_string().into(), request_id).into(),
                     (Instant::now(), pending_response),
                 );
-                debug_assert!(prev_req_id.is_none(), "Expect request id to be unique.");
+                debug_assert!(
+                    prev_req_id.is_none(),
+                    "Expect request id to be unique for new peer."
+                );
```

Comment:
@nazar-pc can you check this is the correct logic for your new APIs?

Reply:
Why is there two branches now rather than just the second one?

Reply:
Are the addresses ignored if the target is already connected?

Reply:
Fixed in commit [60e1465](https://github.com/autonomys/subspace/pull/3420/commits/60e1465ef579e474779034bfc400b13f412937b6)